### PR TITLE
feat(upload): provider upload with command line interface

### LINKED ISSUE #101: clarification on uploading a provider
1. can we make an upload of a module/provider on a distant registry not local one ? 
2. I see that uploading a provider is not supported yet using CLI , what dev work should be done on that ? is there any blockers ?

### PR DESCRIPTION
**Pull request by oliviermichaelis**
This PR aims to fix #101 by adding support for provider upload to the CLI.

The minimal command to upload the provider release artifacts is:
  ```bash
  boring-registry upload provider \
  --storage-s3-bucket <bucket_name> \
  --namespace <namespace> \
  --filename-sha256sums /absolute/path/to/terraform-provider-<name>_<version>_SHA256SUMS
  ```
where `--namespace` points to the `<namespace>` in [our internal storage layout](https://github.com/TierMobility/boring-registry#internal-storage-layout).
The `--filename-sha256sums` has to point to the releases' `SHA256SUMS` file. By default, we scan through this file and assume that all the binary zip archives for which checksums exist are located in the same directory as the SHA256SUMS.
If that might not be the case (for whatever reason), it's possible to use the `filenames-provider-archives` flag to provide the paths to the binary archives in a string array instead. 

Tested on S3 and MinIO.

### CODE REVIEW

File: cmd/upload.go [hunk 1]
```diff
@@ -86,3 +118,174 @@ func uploadModule(cmd *cobra.Command, args []string) error {
...
+
+func validateShaSums(sums *core.Sha256Sums) error {
+	// Check whether the user has given archive paths to upload on the command line as flags.
+	// If not, we try to determine the locations of the provider zip archives based on the path of the *_SHA256SUMS file and the filenames in that file
+	if len(flagProviderArchivePaths) != 0 {
+		if len(sums.Entries) != len(flagProviderArchivePaths) {
+			return fmt.Errorf("the number of provided archive paths doesn't match the number of entries in %s", flagFileSha256Sums)
+		}
+
+	OUTER:
```
Comment (heimweh) [inline cmd/upload.go:229]: Can we change this part to avoid using labeling?

File: pkg/core/provider.go [hunk 2]
```diff
@@ -85,10 +94,38 @@ func NewProviderFromArchive(filename string) (Provider, error) {
...
+	for _, key := range s.GPGPublicKeys {
+		keyring, err := openpgp.ReadArmoredKeyRing(strings.NewReader(key.ASCIIArmor))
+		if err != nil {
+			return fmt.Errorf("error reading signing key: %w", err)
+		}
+
+		_, err = openpgp.CheckDetachedSignature(keyring, bytes.NewReader(sha256Sums), bytes.NewReader(sha256SumsSig), nil)
+
+		// If the signature issuer does not match the key, keep trying the rest of the provided keys.
+		if errors.Is(err, openpgpErrors.ErrUnknownIssuer) {
```
Comment (heimweh) [inline pkg/core/provider.go:115]: Do we want to surface this error or log this as an error?

File: pkg/core/provider.go [hunk 3]
```diff
@@ -120,3 +157,80 @@ type providerOption struct {
...
+	Entries  []Sha256SumsEntry
+	Filename string
+}
+
+// Name returns the name of the provider of the SHA256SUMS file
+func (s *Sha256Sums) Name() string {
+	// RegEx could fail in rare cases as the first capture group doesn't try to match as much as possible
+	r := regexp.MustCompile("^terraform-provider-(?P<name>.+)_(?P<version>.+)_SHA256SUMS$")
+	matches := r.FindStringSubmatch(s.Filename)
+	return matches[1]
```
Comment (heimweh) [inline pkg/core/provider.go:200]: Do we want to check the length of matches as a precaution or check against nil? https://pkg.go.dev/regexp#Regexp.FindStringSubmatch

**Review by heimweh — COMMENTED** [review]
Looks great! Left a few inline comments.

Can't wait to test this :star_struck:

File: pkg/core/provider.go (hunk 2 above)
Comment (oliviermichaelis) [reply pkg/core/provider.go:115]: If it's an `openpgpErrors.ErrUnknownIssuer`, we do not want to return it, as it's an expected error :)

File: cmd/upload.go (hunk 1 above)
Comment (oliviermichaelis) [reply cmd/upload.go]: I don't know why I chose the complicated route when I could've used `break` instead. Fixed this in 5c55979

File: pkg/core/provider.go (hunk 3 above)
Comment (oliviermichaelis) [reply pkg/core/provider.go]: Good point, I addressed this in  5c55979

**Pull request merged**